### PR TITLE
feat: appsmithctl backup error logging and mailer

### PR DESCRIPTION
**Pull request by sum35h**
## Description
- Added error logging through a new module `logger.js` which stores backup logs at `/appsmith-stacks/logs/backup/error.log`
- Added a mailer module `mailer.js` which uses the appsmith mailer smtp credentials from the `docker.env` file to send error mails to admins.
  The error mailer is disabled by default and can be enabled using the `--error-mail `arg on the `backup` command:
  Ex: `appsmithctl backup --error-mail`

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: deploy/docker/utils/bin/constants.js [hunk 1]
```diff
@@ -5,8 +5,14 @@ const RESTORE_PATH = "/appsmith-stacks/data/restore"
...
+const BACKUP_ERROR_LOG_PATH = '/appsmith-stacks/logs/backup'
+
+const MIN_REQUIRED_DISK_SPACE_IN_BYTES = 5368709120 // 5GB
+
 module.exports = {
     BACKUP_PATH,
     RESTORE_PATH,
     DUMP_FILE_NAME,
+    BACKUP_ERROR_LOG_PATH,
+    MIN_REQUIRED_DISK_SPACE_IN_BYTES
```
Comment (sharat87) [inline deploy/docker/utils/bin/constants.js:17]:
```suggestion
    MIN_REQUIRED_DISK_SPACE_IN_BYTES,
```

File: deploy/docker/utils/bin/backup.js [hunk 2]
```diff
@@ -45,9 +57,14 @@ async function run() {
...
+    await logger.backup_error(err.stack);
+    
+    if (command_args.includes('--error_mail')){
+      await mailer.sendBackupErrorToAdmins(err, timestamp);
+    }   
   } finally {
-    utils.start(['backend', 'rts']);
-    process.exit(errorCode);
+      utils.start(['backend', 'rts']);
+      process.exit(errorCode);
```
Comment (sharat87) [inline deploy/docker/utils/bin/backup.js:67]:
```suggestion
    utils.start(['backend', 'rts']);
    process.exit(errorCode);
```

File: deploy/docker/utils/bin/backup.js [hunk 3]
```diff
@@ -6,9 +6,14 @@ const shell = require('shelljs');
 
 const utils = require('./utils');
 const Constants = require('./constants');
+const logger = require('./logger');
+const mailer = require('./mailer')
```
Comment (sharat87) [inline deploy/docker/utils/bin/backup.js:10]:
```suggestion
const mailer = require('./mailer');
```

File: deploy/docker/utils/bin/backup.js [hunk 4]
```diff
@@ -19,9 +24,16 @@ async function run() {
...
-    utils.stop(['backend', 'rts']);
+      utils.stop(['backend', 'rts']);
+
+      console.log('Available free space at /appsmith-stacks')
+      const availSpaceInBytes =  Number(shell.exec('df --output=avail -B 1 /appsmith-stacks | tail -n 1'))
+      console.log('\n')
+
+      if (availSpaceInBytes < Constants.MIN_REQUIRED_DISK_SPACE_IN_BYTES){
+        throw new Error('Not enough space avaliable at /appsmith-stacks. Please ensure availability of atleast 5GB to backup successfully.')
+      }
```
Comment (sharat87) [inline deploy/docker/utils/bin/backup.js:35]:
```suggestion
    utils.stop(['backend', 'rts']);

    console.log('Available free space at /appsmith-stacks');
    const availSpaceInBytes =  Number(shell.exec('df --output=avail -B 1 /appsmith-stacks | tail -n 1'));
    console.log('\n');

    if (availSpaceInBytes < Constants.MIN_REQUIRED_DISK_SPACE_IN_BYTES){
      throw new Error('Not enough space avaliable at /appsmith-stacks. Please ensure availability of atleast 5GB to backup successfully.');
    }
```

File: deploy/docker/utils/bin/logger.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    try {
+      await fsPromises.access(Constants.BACKUP_ERROR_LOG_PATH);
+    } catch (error) {
+      await fsPromises.mkdir(Constants.BACKUP_ERROR_LOG_PATH)
+    }
+    await fsPromises.appendFile(Constants.BACKUP_ERROR_LOG_PATH + '/error.log', new Date().toISOString() + ' ' + err + '\n');
+  }
+
+module.exports = {
+    backup_error
```
Comment (sharat87) [inline deploy/docker/utils/bin/logger.js:14]:
```suggestion
    backup_error,
```

File: deploy/docker/utils/bin/logger.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+const fsPromises = require('fs/promises');
+const Constants = require('./constants');
+
+async function backup_error(err){
+    try {
+      await fsPromises.access(Constants.BACKUP_ERROR_LOG_PATH);
+    } catch (error) {
+      await fsPromises.mkdir(Constants.BACKUP_ERROR_LOG_PATH)
```
Comment (sharat87) [inline deploy/docker/utils/bin/logger.js:8]:
```suggestion
      await fsPromises.mkdir(Constants.BACKUP_ERROR_LOG_PATH);
```

**Review by sharat87 — CHANGES_REQUESTED** [review]
@sum35h, can you check code formatting once please? Please ensure 2-space indentation, and terminating all statements with a semicolon. I see quite a few changes that aren't consistent in these, and that's making it very hard to follow the code flow and review.

Can you fix this and submit for review again please?

File: deploy/docker/utils/bin/mailer.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            const backupFiles = await utils.listLocalBackupFiles();
+            const lastBackupfile = backupFiles.pop()
+            const lastBackupTimestamp = lastBackupfile.match(/appsmith-backup-(.*)\.tar.gz/)[1]
+            const lastBackupPath = Constants.BACKUP_PATH + '/' + lastBackupfile
+        
+            let domainNameOrIP = process.env.APPSMITH_CUSTOM_DOMAIN 
+            console.log('domain= ' + mailFrom)
+            if (domainNameOrIP === ''){
+                console.log('Host IP:')
+                domainNameOrIP = shell.exec('curl -s ifconfig.me')
```
Comment (sum35h) [inline deploy/docker/utils/bin/mailer.js:39]: @sharat87  Would this be acceptable way to get the ip of the instance?

File: deploy/docker/utils/bin/mailer.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        else {
+            const backupFiles = await utils.listLocalBackupFiles();
+            const lastBackupfile = backupFiles.pop();
+            const lastBackupTimestamp = lastBackupfile.match(/appsmith-backup-(.*)\.tar.gz/)[1]
+            const lastBackupPath = Constants.BACKUP_PATH + '/' + lastBackupfile;
+        
+            const domainName = process.env.APPSMITH_CUSTOM_DOMAIN
+            const instanceName = process.env.APPSMITH_INSTANCE_NAME
+
+            let text = 'Appsmith backup did not complete successfully.\n\n ' +
```
Comment (mohanarpit) [inline deploy/docker/utils/bin/mailer.js:38]: Not urgent. But instead of hard-coding the text, is it possible to use a template? This way, all Appsmith emails will look & feel the same. 

In the Appsmith server code base, some of the mail templates can be found at: `app/server/appsmith-server/src/main/resources/email/`. It'll just look slightly more professional than a plain text mail.

File: deploy/docker/utils/bin/mailer.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            const transporter = nodemailer.createTransport({
+                host: mailHost,
+                port: mailPort,
+                auth: {
+                    user: mailUser,
+                    pass: mailPass
+                }
+            });
+           
+            //   console.log('Mail body: ' + text)
```
Comment (mohanarpit) [inline deploy/docker/utils/bin/mailer.js:60]: Suggestion to please remove commented code.

Comment (mohanarpit) [inline deploy/docker/utils/bin/mailer.js:60]: Suggestion to please remove commented code.

File: deploy/docker/utils/bin/mailer.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                await transporter.sendMail({
+                    from: mailFrom,
+                    to: mailTo,
+                    subject: '[Appsmith] ERROR: Backup Failed',
+                    text: text
+                    });
+          }
+        } catch(err){
+            console.log(err);
```
Comment (mohanarpit) [inline deploy/docker/utils/bin/mailer.js:69]: If we are using logger.backup_error, do we need the console.log as well?

Comment (sum35h) [reply deploy/docker/utils/bin/mailer.js:69]: logger.backup_error() only writes to the backup log file, so console.log is being used to display it on terminal.

File: deploy/docker/utils/bin/mailer.js (hunk 8 above)
Comment (sum35h) [reply deploy/docker/utils/bin/mailer.js:38]: Okay, I'll look into this and maybe pick it up after this pr.

File: deploy/docker/utils/bin/backup.js [hunk 11]
```diff
@@ -36,15 +48,18 @@ async function run() {
...
 
     console.log('Finished taking a baceup at', archivePath);
     // console.log('Please remember to also take the `docker.env` separately since it includes sensitive, but critical information.')
 
   } catch (err) {
-    console.log(err);
     errorCode = 1;
+    await logger.backup_error(err.stack);
 
+    if (command_args.includes('--error_mail')) {
```
Comment (sharat87) [inline deploy/docker/utils/bin/backup.js:60]: Can we use `--error-mail` instead of `--error_mail` here, to stay compliant with POSIX style arguments?

File: deploy/docker/utils/bin/backup.js [hunk 12]
```diff
@@ -21,7 +26,14 @@ async function run() {
 
     utils.stop(['backend', 'rts']);
 
-    const timestamp = new Date().toISOString().replace(/:/g, '-')
+    console.log('Available free space at /appsmith-stacks');
+    const availSpaceInBytes = Number(shell.exec('df --output=avail -B 1 /appsmith-stacks | tail -n 1'));
```
Comment (sharat87) [inline deploy/docker/utils/bin/backup.js:30]: Minor, it's recommended to use `parseInt` to turn strings into Number in JS, unless if there's a specific reason not to.

```suggestion
    const availSpaceInBytes = parseInt(shell.exec('df --output=avail -B 1 /appsmith-stacks | tail -n 1'), 10);
```

**Review by sharat87 — CHANGES_REQUESTED** [review]
Looks good. Couple of minor things, let's merge after these two.

File: deploy/docker/utils/bin/mailer.js (hunk 8 above)
Comment (sum35h) [reply deploy/docker/utils/bin/mailer.js]: Created issue for this at [zenhub](https://app.zenhub.com/workspaces/devops-pod-629d9a5ad45779001f66d8e0/issues/appsmithorg/appsmith/14984).

**Review by sharat87 — APPROVED** [review]
